x-pack/plugins/security/server/routes/api/public/roles/put.test.js

```diff
@@ -0,0 +1,502 @@
+
+const application = 'kibana-.kibana';
+
+const createMockServer = () => {
+  const mockServer = new Hapi.Server({ debug: false });
+  mockServer.connection({ port: 8080 });
+  return mockServer;
+};
+
+const defaultPreCheckLicenseImpl = (request, reply) => reply();
+
+const privilegeMap = {
+  'test-kibana-privilege-1': {},
+  'test-kibana-privilege-2': {},
+  'test-kibana-privilege-3': {},
+};
+
+const putRoleTest = (
+  description,
+  { name, payload, preCheckLicenseImpl, callWithRequestImpls = [], asserts }
+) => {
+  test(description, async () => {
+    const mockServer = createMockServer();
+    const mockPreCheckLicense = jest
+      .fn()
+      .mockImplementation(preCheckLicenseImpl);
+    const mockCallWithRequest = jest.fn();
+    for (const impl of callWithRequestImpls) {
+      mockCallWithRequest.mockImplementationOnce(impl);
+    }
+    initPutRolesApi(
+      mockServer,
+      mockCallWithRequest,
+      mockPreCheckLicense,
+      privilegeMap,
+      application,
+    );
+    const headers = {
+      authorization: 'foo',
+    };
+
+    const request = {
+      method: 'PUT',
+      url: `/api/security/role/${name}`,
+      headers,
+      payload,
+    };
+    const { result, statusCode } = await mockServer.inject(request);
+
+    expect(result).toEqual(asserts.result);
+    expect(statusCode).toBe(asserts.statusCode);
+    if (preCheckLicenseImpl) {
+      expect(mockPreCheckLicense).toHaveBeenCalled();
+    } else {
+      expect(mockPreCheckLicense).not.toHaveBeenCalled();
+    }
+    if (asserts.callWithRequests) {
+      for (const args of asserts.callWithRequests) {
+        expect(mockCallWithRequest).toHaveBeenCalledWith(
+          expect.objectContaining({
+            headers: expect.objectContaining({
+              authorization: headers.authorization,
+            }),
+          }),
+          ...args
+        );
+      }
+    } else {
+      expect(mockCallWithRequest).not.toHaveBeenCalled();
+    }
+  });
+};
+
+describe('PUT role', () => {
+  describe('failure', () => {
+    putRoleTest(`requires name in params`, {
+      name: '',
+      payload: {},
+      asserts: {
+        statusCode: 404,
+        result: {
+          error: 'Not Found',
+          statusCode: 404,
+        },
+      },
+    });
+
+    putRoleTest(`requires name in params to not exceed 1024 characters`, {
+      name: 'a'.repeat(1025),
+      payload: {
+        kibana: [
+          {
+            privileges: ['foo']
+          }
+        ]
+      },
+      asserts: {
+        statusCode: 400,
+        result: {
+          error: 'Bad Request',
+          message: `child "name" fails because ["name" length must be less than or equal to 1024 characters long]`,
+          statusCode: 400,
+          validation: {
+            keys: ['name'],
+            source: 'params',
+          },
+        },
+      },
+    });
+
+    putRoleTest(`only allows known Kibana privileges`, {
```

Comment:
This doesn't appear to be testing "only allows known Kibana privileges"

Reply:
`doesn't allow non-known Kibana privilege`?

Reply:
Sorry, that was a really confusing comment. It looks like the test is asserting on `child "name" fails because ["name" length must be less than or equal to 1024 characters long]` instead of on an unknown Kibana privilege.